flow-plugins/flow-maven-plugin/src/main/java/com/vaadin/flow/plugin/maven/GenerateFrontendBOMMojo.java

```
@@ -0,0 +1,87 @@
+import org.apache.maven.plugin.MojoFailureException;
+import org.apache.maven.plugins.annotations.LifecyclePhase;
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.apache.maven.plugins.annotations.ResolutionScope;
+import org.apache.maven.shared.invoker.DefaultInvoker;
+import org.apache.maven.shared.invoker.InvocationRequest;
+import org.apache.maven.shared.invoker.InvocationResult;
+import org.apache.maven.shared.invoker.Invoker;
+import org.apache.maven.shared.invoker.MavenInvocationException;
+
+/**
+ * Goal that generates a CycloneDX SBOM file focused on frontend dependencies.
+ */
+@Mojo(name = "generate-npm-sbom", requiresDependencyResolution = ResolutionScope.COMPILE, defaultPhase = LifecyclePhase.GENERATE_RESOURCES)
+public class GenerateFrontendBOMMojo extends AbstractMojo {
+
+    private static final String OUTPUT_FORMAT = "outputFormat";
+    private static final String OUTPUT_FILE_PATH = "outputDirectory";
+    private static final String PACKAGE_MANIFEST = "packageManifest";
+
+    private static final String GROUP = "org.codehaus.mojo";
+    private static final String ARTIFACT = "exec-maven-plugin";
+    private static final String VERSION = "1.3.2";
+    private static final String GOAL = "exec";
+
+    /**
+     * The CycloneDX output format that should be generated (<code>xml</code>,
+     * <code>json</code> or <code>all</code>).
+     */
+    @Parameter(property = OUTPUT_FORMAT, defaultValue = "json")
+    private String outputFormat;
+
+    /**
+     * The path to the file to be generated.
+     */
+    @Parameter(property = OUTPUT_FILE_PATH, defaultValue = "${project.build.outputDirectory}/resources/bom-npm.json")
+    private String outputFilePath;
+
+    /**
+     * The path to the package.json file to read.
+     */
+    @Parameter(property = PACKAGE_MANIFEST, defaultValue = "./package.json")
+    private String packageManifest;
```

This only works with a package.json if I understand it correctly, ergo it won't work with the "new" modes that is used by flow by default with the development and production bundles. Should it be checked that the file is present?

---

A package.json file is indeed necessary for the generation of the npm sbom. If no such file is present, the goal will fail.
@mshabarov Can you give us your input here?

---

I think the `package.json` file should be checked if it is present. If not, the goal should throw an exception as commented [here](https://github.com/vaadin/flow/pull/18392#issuecomment-1891631609), or log some warning message that it is a requirement to be present.

---

Ok, I agree. Just thinking out loud that in case we have starters or something that run this goal by default, I'd not like it to fail if `package.json` is missing.

---

With the pre-compiled bundles the package.json isn't usually in the project root directory indeed.

But you can get the content of it programmatically using Flow's util method `BundleValidationUtil::getPackageJson`. It will give to the package.json from project root, if it exists, or calculates the dependencies used in application and returns the content of this file to you as a JsonObject.

Also, I think we already discussed a similar problem before with @tamasmak .

There was a conclusion at that point:

```
Maybe even this function won’t be needed. In platform there is already a SBOM [built](https://github.com/vaadin/platform/blob/main/scripts/generateAndCheckSBOM.js#L485) with the same plugin, if we could get that from there that would be the best. In all other cases when the package.json is present we can use it to generate the SBOM.
```

Here is a [discussion in Slack 
](https://vaadin.slack.com/archives/C6X43FE8M/p1692957649050189?thread_ts=1692786162.763219&cid=C6X43FE8M).

---

Yes, I remember on this discussion @mshabarov. The mentioned conclusion is the solution what we use in the AppSec Kit currently. We get the already generated combined (contains the `maven` and `npm` dependencies too) SBOM json file from the platform if the `package.json` is not present, then we parse it and [filter out](https://github.com/vaadin/appsec-kit/blob/main/appsec-kit-backend/src/main/java/com/vaadin/appsec/backend/BillOfMaterialsStore.java#L76) the `maven` dependencies. If we wanted to use the same solution here, then, after filtering out the `maven` dependencies, we would need to save the result to a new SBOM json file.

The purpose of this maven goal would be to help generate a similar SBOM json file but only with `npm` dependencies. We could achieve this with the above solution or with the `BundleValidationUtil::getPackageJson` method in which case we would need to save the `JsonObject` to a `package.json` file, if not present, to be available for the `cyclonedx-npm` plugin and (probably) to delete it after generating the SBOM file, if it wasn't there before.

---

> I think the `package.json` file should be checked if it is present. If not, the goal should throw an exception as commented [here](https://github.com/vaadin/flow/pull/18392#issuecomment-1891631609), or log some warning message that it is a requirement to be present.

We decided to proceed with a warning message that the p`package.json` file is not present for now. Later we can improve the plugin with getting the npm SBOM from elsewhere if needed.

---

> We decided to proceed with a warning message that the `package.json` file is not present for now. Later we can improve the plugin with getting the npm SBOM from elsewhere if needed.

Sorry if being a little thick here, but earlier [David commented](https://github.com/vaadin/flow/pull/18392#issuecomment-1891524586) that "AppSec Kit can handle frontend dependencies even if there's no package.json file." - so why do we care about the package.json file?

---

Because the Maven goal will still fail without a `package.json`. The Maven goal has nothing to do with AppSec Kit. @tamasmak  meant we will update the Maven plugin goal in a future PR so that it can work by doing what the [AppSec Kit does now to get the frontend dependencies](https://github.com/vaadin/flow/pull/18392#discussion_r1452308048).

---

> The Maven goal has nothing to do with AppSec Kit.
Thanks! This clears it up.